neuralmonkey/decoders/sequence_decoder.py

```diff
@@ -0,0 +1,258 @@
+"""Abstract class for decoding sequences left-to-right.
+
+Either for the recurrent decoder, or for the transformer decoder.
+
+The sequence decoder uses the while loop to get the outputs. Descendants should
+only specify the initial state and the while loop body.
+"""
+import math
+from typing import (NamedTuple, Union, Callable, Tuple, cast, Iterable, Type,
+                    List, Optional, Any)
+
+import numpy as np
+import tensorflow as tf
+
+from neuralmonkey.dataset import Dataset
+from neuralmonkey.decorators import tensor
+from neuralmonkey.logging import log
+from neuralmonkey.model.model_part import ModelPart, FeedDict
+from neuralmonkey.nn.utils import dropout
+from neuralmonkey.vocabulary import Vocabulary, START_TOKEN
+
+
+def extend_namedtuple(name: str, parent: Type,
+                      fields: List[Tuple[str, Type]]) -> Type:
+    """Extend a named tuple to contain more elements."""
+    # pylint: disable=protected-access
+    ext_fields = [(k, parent._field_types[k]) for k in parent._fields] + fields
+    # pylint: enable=protected-access
+    return cast(Type, NamedTuple(name, ext_fields))
+
+
+# The LoopState is a structure that works with the tf.while_loop function
+# the decoder loop state stores all the information that is not invariant
+# for the decoder run.
+# pylint: disable=invalid-name
+DecoderHistories = NamedTuple(
+    "DecoderHistories",
+    [("logits", tf.TensorArray),
+     ("decoder_outputs", tf.TensorArray),
+     ("mask", tf.TensorArray)])  # float matrix, 0s and 1s
+
+DecoderConstants = NamedTuple(
+    "DecoderConstants",
+    [("train_inputs", Optional[tf.Tensor])])
+
+DecoderFeedables = NamedTuple(
+    "DecoderFeedables",
+    [("step", tf.Tensor),  # 1D int, number of the step
+     ("finished", tf.Tensor),  # batch-sized, bool
+     ("input_symbol", tf.Tensor),
+     ("prev_logits", tf.Tensor)])
+
+LoopState = NamedTuple(
+    "LoopState",
+    [("histories", Any),
+     ("constants", Any),
+     ("feedables", Any)])
+# pylint: enable=invalid-name
+
+
+# pylint: disable=too-many-public-methods
+# TODO More refactoring needed.. (?)
+class SequenceDecoder(ModelPart):
+
+    def __init__(self,
+                 name: str,
+                 vocabulary: Vocabulary,
+                 data_id: str,
+                 max_output_len: int,
+                 dropout_keep_prob: float = 1.0,
+                 save_checkpoint: str = None,
+                 load_checkpoint: str = None) -> None:
+        ModelPart.__init__(self, name, save_checkpoint, load_checkpoint)
+        log("Initializing decoder, name: '{}'".format(name))
+
+        self.vocabulary = vocabulary
+        self.data_id = data_id
+        self.max_output_len = max_output_len
+        self.dropout_keep_prob = dropout_keep_prob
+
+        # TODO check the values of the parameters (vocab, ...)
```

Comment:
check the values of the parameters (vocab, ...)

Reply:
Pridal jsem check_argument_types() odstranil jsem comment (krome typu argumentu tady imho neni co checkovat).

Reply:
Pridal jsem i docstring k __init__() tak se na to mrkni.

Reply:
neee, check_argument_types tu nemaj bejt - nevolá se to přímo z konfiguráku. Má tu bejt check, že dropout_prob je v (0,1], že max_output_len je kladný, a tak dál...

Reply:
Mas to tam. U vocabulary me nenapada, co by se dalo checkovat.